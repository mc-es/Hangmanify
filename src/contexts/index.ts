import { I18nProvider, useI18n } from './I18nContext';
import { ThemeProvider, useTheme } from './ThemeContext';

export { I18nProvider, ThemeProvider, useI18n, useTheme };
