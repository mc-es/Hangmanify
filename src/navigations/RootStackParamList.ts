import {
  type NavigationProp,
  type RouteProp,
  useNavigation as useNativeNavigation,
  useRoute as useNativeRoute,
} from '@react-navigation/native';

export const enum RouteNames {
  ABOUT = 'About',
  HOME = 'Home',
}

export type RootStackParamList = {
  [RouteNames.ABOUT]: { name: string };
  [RouteNames.HOME]: undefined;
};

/**
 * ### Returns the typed navigation object for a given route
 * @template T
 * @returns {NavigationProp<RootStackParamList, T>} Strongly typed navigation object for screen `T`
 * @example
 * // File: Home.tsx
 * import {useNavigation, RouteNames} from './path/to/navigations'
 * const navigation = useNavigation<RouteNames.HOME>(); // Home screen navigation
 * navigation.navigate(RouteNames.PROFILE, { userId: '123' });
 */
function useNavigation(): never;
function useNavigation<T extends keyof RootStackParamList>(): NavigationProp<
  RootStackParamList,
  T
>;
function useNavigation<T extends keyof RootStackParamList>(): NavigationProp<
  RootStackParamList,
  T
> {
  return useNativeNavigation<NavigationProp<RootStackParamList, T>>();
}
/**
 * ### Returns the typed route object for a given screen
 * @template T
 * @returns {RouteProp<RootStackParamList, T>} Strongly typed route object for screen `T`
 * @example
 * // File: Profile.tsx
 * import {useRoute, RouteNames} from './path/to/navigations'
 * const route = useRoute<RouteNames.PROFILE>(); // Profile screen route
 * console.log(route.params.userId);
 */
function useRoute(): never;
function useRoute<T extends keyof RootStackParamList>(): RouteProp<RootStackParamList, T>;
function useRoute<T extends keyof RootStackParamList>(): RouteProp<
  RootStackParamList,
  T
> {
  return useNativeRoute<RouteProp<RootStackParamList, T>>();
}

export { useNavigation, useRoute };
