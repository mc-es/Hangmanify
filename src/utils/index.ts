import { cn } from './cn';
import { dimensions } from './dimensions';
import { helpers } from './helpers';

export { cn, dimensions, helpers };
