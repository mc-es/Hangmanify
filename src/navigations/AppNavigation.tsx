import React from 'react';
import { SafeAreaProvider } from 'react-native-safe-area-context';

import { NavigationContainer } from '@react-navigation/native';
import { createNativeStackNavigator } from '@react-navigation/native-stack';

import About from 'src/screens/About';
import Home from 'src/screens/Home';

import { type RootStackParamList, RouteNames } from './RootStackParamList';

const Stack = createNativeStackNavigator<RootStackParamList>();

const AppNavigation = (): React.JSX.Element => (
  <SafeAreaProvider>
    <NavigationContainer>
      <Stack.Navigator initialRouteName={RouteNames.HOME} screenOptions={{ headerShown: false }}>
        <Stack.Screen component={Home} name={RouteNames.HOME} />
        <Stack.Screen component={About} name={RouteNames.ABOUT} />
      </Stack.Navigator>
    </NavigationContainer>
  </SafeAreaProvider>
);

export default AppNavigation;
