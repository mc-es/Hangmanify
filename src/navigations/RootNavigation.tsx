import React from 'react';
import { SafeAreaProvider } from 'react-native-safe-area-context';

import { NavigationContainer } from '@react-navigation/native';
import { createNativeStackNavigator } from '@react-navigation/native-stack';

import About from 'src/screens/About';
import Home from 'src/screens/Home';

import { RootNavigations, type RootStackParamList } from './RootStackParamList';

const Stack = createNativeStackNavigator<RootStackParamList>();

const RootNavigation = (): React.JSX.Element => (
  <SafeAreaProvider>
    <NavigationContainer>
      <Stack.Navigator
        initialRouteName={RootNavigations.HOME}
        screenOptions={{ headerShown: false }}
      >
        <Stack.Screen component={Home} name={RootNavigations.HOME} />
        <Stack.Screen component={About} name={RootNavigations.ABOUT} />
      </Stack.Navigator>
    </NavigationContainer>
  </SafeAreaProvider>
);

export default RootNavigation;
