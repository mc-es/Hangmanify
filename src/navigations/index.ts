import { RouteNames, useNavigation, useRoute } from './RootStackParamList';

export { RouteNames, useNavigation, useRoute };
