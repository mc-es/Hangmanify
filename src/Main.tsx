import React from 'react';
import { StyleSheet } from 'react-native';
import { SafeAreaView } from 'react-native-safe-area-context';

import { StatusBar } from 'expo-status-bar';

// don't import from '../navigations' here â€” it causes a circular dependency:
import AppNavigation from './navigations/AppNavigation';

import { useTheme } from './contexts';

const Main = (): React.JSX.Element => {
  const { theme } = useTheme();

  return (
    <SafeAreaView style={[styles.container, { backgroundColor: theme.palette.background }]}>
      <StatusBar backgroundColor={theme.palette.background} style="auto" />
      <AppNavigation />
    </SafeAreaView>
  );
};

export default Main;

const styles = StyleSheet.create({
  container: { flex: 1, width: '100%' },
});
