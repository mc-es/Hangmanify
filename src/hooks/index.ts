/* eslint-disable @custom-typescript/no-direct-hook-imports */
import { useI18n } from 'src/contexts/I18nContext';
import { useTheme } from 'src/contexts/ThemeContext';
import { useNavigation, useRoute } from 'src/navigations/RootStackParamList';
import { useCounter, useGlobalText } from 'src/stores/useStore';

export {
  useNavigation as useAppNavigation,
  useRoute as useAppRoute,
  useTheme as useAppTheme,
  useI18n as useAppI18n,
  useCounter as useAppCounter,
  useGlobalText as useAppGlobalText,
};
