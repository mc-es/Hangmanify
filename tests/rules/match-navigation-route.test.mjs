/**
 * @fileoverview Tests for match-navigation-route rule.
 * @see https://github.com/mces58/Hangmanify/blob/master/guides/rules/match-navigation-route.md
 */
import rule from 'plugins/rules/match-navigation-route.mjs';

import { TEST_FILENAME_PATH } from './Test';
import { runTest } from './utils.js';

const validCases = [
  {
    name: 'passes with correct generic for useNavigation',
    filename: TEST_FILENAME_PATH,
    code: `
      import { useNavigation } from 'src/navigations';
      const nav = useNavigation<RouteNames.TEST>();
    `,
  },
  {
    name: 'passes with correct generic for useRoute',
    filename: TEST_FILENAME_PATH,
    code: `
      import { useRoute, } from 'src/navigations';
      const route = useRoute<RouteNames.TEST>();
    `,
  },
  {
    name: 'passes with correct generics for both hooks',
    filename: TEST_FILENAME_PATH,
    code: `
      import { useNavigation, useRoute } from 'src/navigations';
      const nav = useNavigation<RouteNames.TEST>();
      const route = useRoute<RouteNames.TEST>();
    `,
  },
];

const invalidCases = [
  {
    name: 'fails with mismatched generic in useNavigation',
    filename: TEST_FILENAME_PATH,
    code: `
      import { useNavigation } from 'src/navigations';
      const nav = useNavigation<RouteNames.PROFILE>();
    `,
    expectedErrors: 1,
  },
  {
    name: 'fails with mismatched generic in useRoute',
    filename: TEST_FILENAME_PATH,
    code: `
      import { useRoute } from 'src/navigations';
      const route = useRoute<RouteNames.SETTINGS>();
    `,
    expectedErrors: 1,
  },
  {
    name: 'fails when generics are missing in both hooks',
    filename: TEST_FILENAME_PATH,
    code: `
      import { useNavigation, useRoute } from 'src/navigations';
      const nav = useNavigation();
      const route = useRoute();
    `,
    expectedErrors: 2,
  },
  {
    name: 'fails when importing from @react-navigation/native',
    filename: TEST_FILENAME_PATH,
    code: `import { useNavigation } from '@react-navigation/native';`,
    expectedErrors: 1,
  },
];

runTest('match-navigation-route', rule, validCases, invalidCases);
