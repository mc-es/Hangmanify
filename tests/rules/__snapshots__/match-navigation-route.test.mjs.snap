// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`match-navigation-route rule -> invalid cases -> fails when generics are missing in both hooks 1`] = `
[
  {
    "column": 19,
    "endColumn": 34,
    "endLine": 3,
    "line": 3,
    "message": "You must specify a generic like useNavigation<RouteNames.TEST> in this file.",
    "messageId": "missing",
    "nodeType": "CallExpression",
    "ruleId": "__test_plugin__/match-navigation-route",
    "severity": 2,
  },
  {
    "column": 21,
    "endColumn": 31,
    "endLine": 4,
    "line": 4,
    "message": "You must specify a generic like useRoute<RouteNames.TEST> in this file.",
    "messageId": "missing",
    "nodeType": "CallExpression",
    "ruleId": "__test_plugin__/match-navigation-route",
    "severity": 2,
  },
]
`;

exports[`match-navigation-route rule -> invalid cases -> fails when importing from @react-navigation/native 1`] = `
[
  {
    "column": 10,
    "endColumn": 23,
    "endLine": 1,
    "line": 1,
    "message": "Do not import "useNavigation" from @react-navigation/native. Use your local wrapper instead.",
    "messageId": "bannedImport",
    "nodeType": "ImportSpecifier",
    "ruleId": "__test_plugin__/match-navigation-route",
    "severity": 2,
  },
]
`;

exports[`match-navigation-route rule -> invalid cases -> fails with mismatched generic in useNavigation 1`] = `
[
  {
    "column": 33,
    "endColumn": 51,
    "endLine": 3,
    "line": 3,
    "message": "Generic "RouteNames.PROFILE" does not match expected "RouteNames.TEST" based on filename.",
    "messageId": "mismatch",
    "nodeType": "TSTypeReference",
    "ruleId": "__test_plugin__/match-navigation-route",
    "severity": 2,
  },
]
`;

exports[`match-navigation-route rule -> invalid cases -> fails with mismatched generic in useRoute 1`] = `
[
  {
    "column": 30,
    "endColumn": 49,
    "endLine": 3,
    "line": 3,
    "message": "Generic "RouteNames.SETTINGS" does not match expected "RouteNames.TEST" based on filename.",
    "messageId": "mismatch",
    "nodeType": "TSTypeReference",
    "ruleId": "__test_plugin__/match-navigation-route",
    "severity": 2,
  },
]
`;
